lib/src/constants.dart

```diff
@@ -437,22 +437,22 @@ const LOWER_Z = 122;
 const UPPER_A = 65;
 const UPPER_Z = 90;
 
-bool isLetterOrDigit(String char) => isLetter(char) || isDigit(char);
+bool isLetterOrDigit(String? char) => isLetter(char) || isDigit(char);
 
 // Note: this is intentially ASCII only
-bool isLetter(String char) {
+bool isLetter(String? char) {
   if (char == null) return false;
   final cc = char.codeUnitAt(0);
   return cc >= LOWER_A && cc <= LOWER_Z || cc >= UPPER_A && cc <= UPPER_Z;
 }
 
-bool isDigit(String char) {
+bool isDigit(String? char) {
   if (char == null) return false;
   final cc = char.codeUnitAt(0);
   return cc >= ZERO && cc < ZERO + 10;
 }
 
-bool isHexDigit(String char) {
+bool isHexDigit(String? char) {
```

Comment:
Would make all of these non-nullable. Passing `null` blindly to helper functions is a bad idea in general.

Might not work with some existing code, but they'll just have to fix that before migrating to null safety. It should keep running for legacy code.

Reply:
> Might not work with some existing code, but they'll just have to fix that before migrating to null safety. It should keep running for legacy code.

None of these should be public outside this package. The problem is with all the noise within this package.

```diff
diff --git a/lib/src/constants.dart b/lib/src/constants.dart
index 52c08d7..d08867f 100644
--- a/lib/src/constants.dart
+++ b/lib/src/constants.dart
@@ -437,23 +437,20 @@ const LOWER_Z = 122;
 const UPPER_A = 65;
 const UPPER_Z = 90;
 
-bool isLetterOrDigit(String? char) => isLetter(char) || isDigit(char);
+bool isLetterOrDigit(String char) => isLetter(char) || isDigit(char);
 
 // Note: this is intentially ASCII only
-bool isLetter(String? char) {
-  if (char == null) return false;
+bool isLetter(String char) {
   final cc = char.codeUnitAt(0);
   return cc >= LOWER_A && cc <= LOWER_Z || cc >= UPPER_A && cc <= UPPER_Z;
 }
 
-bool isDigit(String? char) {
-  if (char == null) return false;
+bool isDigit(String char) {
   final cc = char.codeUnitAt(0);
   return cc >= ZERO && cc < ZERO + 10;
 }
 
-bool isHexDigit(String? char) {
-  if (char == null) return false;
+bool isHexDigit(String char) {
   switch (char.codeUnitAt(0)) {
     case 48:
     case 49:
diff --git a/lib/src/tokenizer.dart b/lib/src/tokenizer.dart
index acd388f..a180cbd 100644
--- a/lib/src/tokenizer.dart
+++ b/lib/src/tokenizer.dart
@@ -179,7 +179,7 @@ class HtmlTokenizer implements Iterator<Token> {
     // Consume all the characters that are in range while making sure we
     // don't hit an EOF.
     var c = stream.char();
-    while (allowed(c) && c != eof) {
+    while (c != null && allowed(c)) {
       charStack.add(c);
       c = stream.char();
     }
@@ -276,8 +276,9 @@ class HtmlTokenizer implements Iterator<Token> {
       }
 
       // charStack.last should be the first digit
-      if (hex && isHexDigit(charStack.last) ||
-          (!hex && isDigit(charStack.last))) {
+      final last = charStack.last;
+      if (hex && last != null && isHexDigit(last) ||
+          (!hex && last != null && isDigit(last))) {
         // At least one digit found, so consume the whole number
         stream.unget(charStack.last);
         output = consumeNumberEntity(hex);
@@ -329,7 +330,8 @@ class HtmlTokenizer implements Iterator<Token> {
         }
         if (lastChar != ';' &&
             fromAttribute &&
-            (isLetterOrDigit(charStack[entityLen]) ||
+            charStack[entityLen] != null &&
+            (isLetterOrDigit(charStack[entityLen]!) ||
                 charStack[entityLen] == '=')) {
           stream.unget(charStack.removeLast());
           output = '&${charStack.join()}';
@@ -518,7 +520,7 @@ class HtmlTokenizer implements Iterator<Token> {
       state = markupDeclarationOpenState;
     } else if (data == '/') {
       state = closeTagOpenState;
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       currentToken = StartTagToken(data);
       state = tagNameState;
     } else if (data == '>') {
@@ -545,7 +547,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool closeTagOpenState() {
     final data = stream.char();
-    if (isLetter(data)) {
+    if (data != null && isLetter(data)) {
       currentToken = EndTagToken(data);
       state = tagNameState;
     } else if (data == '>') {
@@ -602,7 +604,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool rcdataEndTagOpenState() {
     final data = stream.char();
-    if (isLetter(data)) {
+    if (data != null && isLetter(data)) {
       _buffer.write(data);
       state = rcdataEndTagNameState;
     } else {
@@ -632,7 +634,7 @@ class HtmlTokenizer implements Iterator<Token> {
       currentToken = EndTagToken('$_buffer');
       emitCurrentToken();
       state = dataState;
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _buffer.write(data);
     } else {
       _addToken(CharactersToken('</$_buffer'));
@@ -657,7 +659,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool rawtextEndTagOpenState() {
     final data = stream.char();
-    if (isLetter(data)) {
+    if (data != null && isLetter(data)) {
       _buffer.write(data);
       state = rawtextEndTagNameState;
     } else {
@@ -681,7 +683,7 @@ class HtmlTokenizer implements Iterator<Token> {
       currentToken = EndTagToken('$_buffer');
       emitCurrentToken();
       state = dataState;
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _buffer.write(data);
     } else {
       _addToken(CharactersToken('</$_buffer'));
@@ -709,7 +711,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool scriptDataEndTagOpenState() {
     final data = stream.char();
-    if (isLetter(data)) {
+    if (data != null && isLetter(data)) {
       _buffer.write(data);
       state = scriptDataEndTagNameState;
     } else {
@@ -733,7 +735,7 @@ class HtmlTokenizer implements Iterator<Token> {
       currentToken = EndTagToken('$_buffer');
       emitCurrentToken();
       state = dataState;
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _buffer.write(data);
     } else {
       _addToken(CharactersToken('</$_buffer'));
@@ -833,7 +835,7 @@ class HtmlTokenizer implements Iterator<Token> {
     if (data == '/') {
       _buffer.clear();
       state = scriptDataEscapedEndTagOpenState;
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _addToken(CharactersToken('<$data'));
       _buffer.clear();
       _buffer.write(data);
@@ -848,7 +850,7 @@ class HtmlTokenizer implements Iterator<Token> {
 
   bool scriptDataEscapedEndTagOpenState() {
     final data = stream.char();
-    if (isLetter(data)) {
+    if (data != null && isLetter(data)) {
       _buffer.clear();
       _buffer.write(data);
       state = scriptDataEscapedEndTagNameState;
@@ -873,7 +875,7 @@ class HtmlTokenizer implements Iterator<Token> {
       currentToken = EndTagToken('$_buffer');
       emitCurrentToken();
       state = dataState;
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _buffer.write(data);
     } else {
       _addToken(CharactersToken('</$_buffer'));
@@ -892,7 +894,7 @@ class HtmlTokenizer implements Iterator<Token> {
       } else {
         state = scriptDataEscapedState;
       }
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _addToken(CharactersToken(data));
       _buffer.write(data);
     } else {
@@ -992,7 +994,7 @@ class HtmlTokenizer implements Iterator<Token> {
       } else {
         state = scriptDataDoubleEscapedState;
       }
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _addToken(CharactersToken(data));
       _buffer.write(data);
     } else {
@@ -1006,7 +1008,7 @@ class HtmlTokenizer implements Iterator<Token> {
     final data = stream.char();
     if (isWhitespace(data)) {
       stream.charsUntil(spaceCharacters, true);
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _addAttribute(data);
       state = attributeNameState;
     } else if (data == '>') {
@@ -1037,7 +1039,7 @@ class HtmlTokenizer implements Iterator<Token> {
     var emitToken = false;
     if (data == '=') {
       state = beforeAttributeValueState;
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _attributeName.write(data);
       _attributeName.write(stream.charsUntil(asciiLetters, true));
       leavingThisState = false;
@@ -1099,7 +1101,7 @@ class HtmlTokenizer implements Iterator<Token> {
       state = beforeAttributeValueState;
     } else if (data == '>') {
       emitCurrentToken();
-    } else if (isLetter(data)) {
+    } else if (data != null && isLetter(data)) {
       _addAttribute(data);
       state = attributeNameState;
     } else if (data == '/') {
```